File: QuickSort/QuickSort.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <algorithm>
#include <array>

std::ostream &operator<<(std::ostream &out, const std::vector<int> &v) {
    for (auto e : v) out << e << ' ';
    return out;
}

std::vector<int> readInput(std::string fileName) {
    std::ifstream in(fileName);
    std::vector<int> v;
    int n;
    while (in >> n) v.push_back(n);
    return v;
}

struct PivotFirst {
    int operator()(const std::vector<int> &v, int l, int r) { return l; }
};

struct PivotLast {
    int operator()(const std::vector<int> &v, int l, int r) { return r; }
};

struct PivotMedian {
    int operator()(const std::vector<int> &v, int l, int r) {
        int m = (l + r) / 2;
        auto c = std::array<std::pair<int, int>, 3>{
            std::pair<int, int>{l, v[l]},
            std::pair<int, int>{m, v[m]},
            std::pair<int, int>{r, v[r]} };
        std::sort(c.begin(), c.end(), [](auto a, auto b) { return a.second < b.second; });
        return c[1].first;
    }
};

template <typename PivotFn>
int quicksort(std::vector<int> &v, int l, int r) {
    const int window = r - l + 1;
    if (window <= 1) return std::max(0, window - 1);
    const int pivot = PivotFn()(v, l, r);
    if (pivot != l) std::swap(v[pivot], v[l]);
    int feed = l + 1;
    for (int i = l + 1; i <= r; ++i) {
        if (v[i] < v[l]) std::swap(v[feed++], v[i]);
    }
    std::swap(v[l], v[feed - 1]);
    return std::max(0, window - 1) + quicksort<PivotFn>(v, l, feed - 2) + quicksort<PivotFn>(v, feed, r);
}

int main(int argc, char *argv[]) {
    if (argc < 2) return 1;
    const auto input = readInput(argv[1]);

    {
        auto work = input;
        std::cout << "PivotFirst " << quicksort<PivotFirst>(work, 0, work.size() - 1) << std::endl;
        std::cout << "is_sorted " << std::is_sorted(work.begin(), work.end()) << std::endl;
    }

    {
        auto work = input;
        std::cout << "PivotLast " << quicksort<PivotLast>(work, 0, work.size() - 1) << std::endl;
        std::cout << "is_sorted " << std::is_sorted(work.begin(), work.end()) << std::endl;
    }

    {
        auto work = input;
        std::cout << "PivotMedian " << quicksort<PivotMedian>(work, 0, work.size() - 1) << std::endl;
        std::cout << "is_sorted " << std::is_sorted(work.begin(), work.end()) << std::endl;
    }
}

File: QuickSort/QuickSort.hs
import System.Environment (getArgs)
import Control.Exception (assert)
import Control.Monad.ST (runST, ST)
import Control.Monad (when, forM_, when)
import Data.STRef

import qualified Data.Vector.Unboxed.Mutable as MV
import qualified Data.Vector.Unboxed as V
import Data.List (sortBy)
import Data.Function (on)

readInput :: String -> IO (V.Vector Int)
readInput file = V.fromList . map read . filter (not . null) . lines <$> readFile file

type FindPivot s a = V.Vector a -> Int -> Int -> Int

pivotLeft :: (Ord a) => FindPivot s a
pivotLeft _ l _ = l

pivotRight :: (Ord a) => FindPivot s a
pivotRight _ _ r = r

pivotMedian :: (Ord a, V.Unbox a) => FindPivot s a
pivotMedian v l r = (fst . head . tail) $ sortBy (compare `on` snd) (zip arr (map (v V.!) arr))
    where arr = [l, r, (l + r) `div` 2]

partition :: (Ord a, V.Unbox a) => FindPivot s a -> MV.MVector s a -> Int -> Int -> ST s Int
partition p v l r = do
    pivotInput_ <- V.freeze v
    let pivot = p pivotInput_ l r
    when (pivot /= l) $ MV.swap v pivot l
    feed <- newSTRef (l + 1)
    forM_ [l + 1 .. r] $ \i -> do
        lcomp <- MV.read v i
        rcomp <- MV.read v l
        when (lcomp < rcomp) $ do
            feed_ <- readSTRef feed
            MV.swap v feed_ i
            writeSTRef feed (feed_ + 1)
    feed_ <- readSTRef feed
    MV.swap v l (feed_ - 1)
    return feed_

type Comparisons = Int

quickSort_ :: (Ord a, V.Unbox a) => FindPivot s a -> MV.MVector s a -> Int -> Int -> ST s Comparisons
quickSort_ p v l r = do
    let window = r - l + 1
        numCompares = max 0 (window - 1)
    if window <= 1 then
        return numCompares
    else do
        feed <- partition p v l r
        leftCompares <- quickSort_ p v l (feed - 2)
        rightCompares <- quickSort_ p v feed r
        return (numCompares + leftCompares + rightCompares)

quickSort :: (Ord a, V.Unbox a) => FindPivot s a -> V.Vector a -> (V.Vector a, Comparisons)
quickSort p v = runST $ do
    work_ <- V.thaw v
    comps <- quickSort_ p work_ 0 (V.length v - 1)
    sorted <- V.freeze work_
    return (sorted, comps)

main :: IO ()
main = do
    argFile <- (\x -> if null x then "QuickSort/input.txt" else head x) <$> getArgs
    input <- readInput argFile
    putStrLn ((("PivotLeft " ++) . show . snd . quickSort pivotLeft) input)
    putStrLn ((("PivotRight " ++) . show . snd . quickSort pivotRight) input)
    putStrLn ((("PivotMedian " ++) . show . snd . quickSort pivotMedian) input)
    return ()
